KAFKA-7225: Corrected system tests by generating external properties file

rhauch opened the pull request:
Fix system tests from earlier #5445 by moving to the `ConnectSystemBase` class the creation & cleanup of a file that can be used as externalized secrets in connector configs. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-7225: Corrected system tests by generating external properties file

rhauch commented:
First run of system tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1907/

wicknicks commented on tests/kafkatest/services/connect.py:
```diff
@@ -145,7 +158,7 @@ def restart_node(self, node, clean_shutdown=True):
         node.account.kill_process("connect", clean_shutdown=False, allow_fail=True)
         self.security_config.clean_node(node)
-        all_files = " ".join([self.CONFIG_FILE, self.LOG4J_CONFIG_FILE, self.PID_FILE, self.LOG_FILE, self.STDOUT_FILE, self.STDERR_FILE] + self.config_filenames() + self.files)
+        all_files = " ".join([self.CONFIG_FILE, self.LOG4J_CONFIG_FILE, self.PID_FILE, self.LOG_FILE, self.STDOUT_FILE, self.STDERR_FILE, self.EXTERNAL_FILE] + self.config_filenames() + self.files)
```
should we delete the EXTERNAL file only if it was created by the `self.external_config_template_func` flag? I think the rm command will fail if the file doesn't exist.

rhauch replied:
At least on MacOS, `rm -rf x` does not fail if the file doesn't exist, but it does fail if it exists but can't be deleted.

rhauch replied:
I also checked the debug build output of multiple tests, including those that did and did not generate this file. No problems removing the file.

wicknicks replied:
makes sense. Thanks, @rhauch!

wicknicks reviewed (commented):
Thanks, @rhauch! One comment.

rhauch commented:
@wicknicks, @rayokota: The [branch build](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1907/) passed with flying colors. Thoughts?

@kkonstantine, @ewencp: care to take a pass at these changes? Hopefully they're idiomatic. :-)

rayokota commented:
LGTM, thanks for generalizing this!

wicknicks reviewed (approved):
Thanks, @rhauch!

rhauch commented:
@ijuma, @ewencp, or @hachikuji: would you mind reviewing and if acceptable and merging back as far as `2.0.x`? This will fix one of the AK system test failures.

kkonstantine reviewed (approved):
LGTM

rhauch commented:
@ewencp, @ijuma, @hachikuji: ping again for a review & merge back as far as `2.0.x` to fix AK system test failures.

hachikuji commented on tests/kafkatest/tests/connect/connect_test.py:
```diff
@@ -100,14 +100,12 @@ def test_file_source_and_sink(self, converter="org.apache.kafka.connect.json.Jso
-        self.sink.node.account.create_file(sink_external_props, self.render('connect-file-external.properties'))
         self.sink.set_configs(lambda node: self.render("connect-standalone.properties", node=node), [self.render("connect-file-sink.properties")])
 
+        self.source.set_external_configs(lambda node: self.render("connect-file-external.properties", node=node))
```
One thing that confuses me a little bit is the dependence on the name "connect-file-external.properties." It is hard-coded above as `EXTERNAL_FILE`, but then why do we need it again here? I'm actually wondering if the full lambda generalization is needed. Could this instead be a flag which simply indicates whether rendering of the external config is desired or not? Or perhaps you had additional use cases in mind?

rhauch replied:
The files happen to be the same, but this actually refers to the test-specific template whereas `EXTERNAL_FILE` refers to the generated file used by the connector. I'll rename the test-specific template file so it's less confusing.

rhauch replied:
To be honest, I used a lambda to follow the same pattern as the existing `set_configs` method.

hachikuji replied:
Ack. Makes sense now.

rhauch committed:
KAFKA-7225: Changed name of generated external config file

Changed the name of the external configuration file that can be generated by tests and optionally used in connector configurations. The template that the `connect.py` standalone tests use to generate the file is now a different.

rhauch commented:
Additional run of system tests with the recent commit: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1924/

rhauch committed:
KAFKA-7225: Correct reference to external config file

rhauch commented:
Run of Connect system tests with the most recent correction: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1927/

hachikuji reviewed (approved):
LGTM. Thanks for the fix.
